src/sinks/file/mod.rs

```diff
@@ -88,98 +95,143 @@ impl PartitionedFileSink {
-            }
+        trace!(message = "Writing an event to file.", ?path);
+        if let Err(error) = write_event_to_file(file, event, &self.encoding).await {
+            error!(message = "Failed to write file.", ?path, %error);
         }
+    }
+}
 
-        let mut closed_files = Vec::new();
+async fn open_file(path: impl AsRef<std::path::Path>) -> std::io::Result<File> {
+    let parent = path.as_ref().parent();
 
-        for (key, file) in &mut self.closing {
-            if let Async::Ready(()) = file.close().unwrap() {
-                closed_files.push(key.clone());
-            }
-        }
+    if let Some(parent) = parent {
+        fs::create_dir_all(parent).await?;
+    }
 
-        for closed_file in closed_files {
-            self.closing.remove(&closed_file);
-        }
+    fs::OpenOptions::new()
+        .read(false)
+        .write(true)
+        .create(true)
+        .open(path)
+        .await
+}
 
-        // Set `next_linger_timeout` to the oldest file's elapsed time since last
-        // write minus the idle_timeout.
-        if let Some(min_last_accessed) = self
-            .last_accessed
-            .iter()
-            .map(|(_, v)| v)
-            .filter(|l| l.elapsed().as_secs() < self.idle_timeout_secs)
-            .min()
-        {
-            let next_timeout = self.idle_timeout_secs - min_last_accessed.elapsed().as_secs();
-            let linger_deadline = *min_last_accessed - std::time::Duration::from_secs(next_timeout);
-            self.next_linger_timeout = Some(Delay::new(linger_deadline));
-        }
+pub fn encode_event(encoding: &Encoding, event: Event) -> Vec<u8> {
+    let log = event.into_log();
+    match encoding {
+        Encoding::Ndjson => serde_json::to_vec(&log).expect("Unable to encode event as JSON."),
+        Encoding::Text => log
+            .get(&event::log_schema().message_key())
+            .map(|v| v.to_string_lossy().into_bytes())
+            .unwrap_or_default(),
+    }
+}
 
-        if let Some(next_linger) = &mut self.next_linger_timeout {
-            next_linger
-                .poll()
-                .expect("This is a bug; we are always in a timer context");
-        }
+async fn write_event_to_file(
+    file: &mut File,
+    event: Event,
+    encoding: &Encoding,
+) -> Result<(), std::io::Error> {
+    let mut buf = encode_event(encoding, event);
+    buf.push(b'\n');
+    file.write_all(&buf[..]).await
+}
 
-        // This sink has completely fnished when one of these is true in order:
-        // 1. There are no active partitions and not files currently closing.
-        // 2. All files have been flushed completely
-        if (self.partitions.is_empty() && self.closing.is_empty()) || all_files_ready {
-            Ok(Async::Ready(()))
-        } else {
-            Ok(Async::NotReady)
-        }
+#[async_trait]
+impl StreamingSink for FileSink {
+    async fn run<'a>(
+        &'a mut self,
+        input: impl Stream<Item = Event> + Send + Sync + 'a,
```

Comment:
this life time should be inferred? Not sure why you need to add this?

Reply:
Oh, it does. I wonder how I ended up with this.

Reply:
This is what's happening if I remove `'a`.

```
error[E0309]: the parameter type `impl Stream<Item = Event> + Send + Sync` may not live long enough
   --> src/sinks/file/mod.rs:233:28
    |
233 |       ) -> crate::Result<()> {
    |  ____________________________^
234 | |         FileSink::run(self, input).await
235 | |     }
    | |_____^
    |
note: ...so that the type `impl core::future::future::Future` will meet its required lifetime bounds
   --> src/sinks/file/mod.rs:233:28
    |
233 |       ) -> crate::Result<()> {
    |  ____________________________^
234 | |         FileSink::run(self, input).await
235 | |     }
    | |_____^
help: consider adding an explicit lifetime bound  `'async_trait` to `impl Stream<Item = Event> + Send + Sync`...
    |
232 |         input: impl Stream<Item = Event> + Send + Sync + 'async_trait,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So, I think it makes sense. The default rules don't work because `async_trait` macro injects it's own `'async_trait` lifetime. In fact, this is sort of ok: we want our lifetime to be of not the future or it's pinned `self`, but of the owning object, and real `self`.